File: app/src/main/java/com/romankaranchuk/musicplayer/presentation/ui/player/PlayerPagerAdapter.java
package com.romankaranchuk.musicplayer.presentation.ui.player;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;

import com.romankaranchuk.musicplayer.data.Song;

import java.util.ArrayList;



public class PlayerPagerAdapter extends FragmentPagerAdapter {
    private ArrayList<Song> mSongs;

    PlayerPagerAdapter(FragmentManager fragmentManager, ArrayList<Song> mSongs){
        super(fragmentManager);
        this.mSongs = mSongs;
    }

    @Override
    public int getCount(){
        return mSongs.size();
    }

    @Override
    public Fragment getItem(int position){
        return PlayerPageFragment.newInstance(position);
    }

}

File: app/src/main/java/com/romankaranchuk/musicplayer/presentation/ui/player/page/PlayerPagerAdapter.kt
package com.romankaranchuk.musicplayer.presentation.ui.player.page

import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DiffUtil
import androidx.viewpager2.adapter.FragmentStateAdapter

import com.romankaranchuk.musicplayer.data.Song

import java.util.ArrayList

class PlayerPagerAdapter(fragment: Fragment) : FragmentStateAdapter(fragment) {

    private val songs = ArrayList<Song>()

    fun updateSongs(newSongs: List<Song>) {
        val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize(): Int {
                return songs.size
            }

            override fun getNewListSize(): Int {
                return newSongs.size
            }

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return songs[oldItemPosition].id == newSongs[newItemPosition].id
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                val oldItem = songs[oldItemPosition]
                val newItem = songs[newItemPosition]
                return oldItem == newItem
            }
        })
        this.songs.clear()
        this.songs.addAll(newSongs)

        diffResult.dispatchUpdatesTo(this)
    }

    override fun getItemCount(): Int {
        return songs.size
    }

    override fun createFragment(position: Int): Fragment {
        return PlayerPageFragment.newInstance(position)
    }
}
